### ISSUE TITLE
Linear Chain CRF的实现细节

### ISSUE THREAD

**Issue by shayuewt** [PaddlePaddle/PaddleNLP#1401]
最近在学习`Linear Chain CRF`，个人也进行了相应的实现。但是自实现的线性链CRF中没有加入[START]以及[END]两个token。后续训练的时候，随着迭代的进行，我发现NLLoss会变为负数。

在网络上搜索相关问题时，发现了你们的版本。注意到在paddleNLP的实现中，将是否引入[START]和[END]这两个token作为一个参数。

开这个issue主要是想请教一下，NLLoss变为负数同`是否加[START]、[END]`相关嘛？麻烦能够指点一二。

**Comment by joey12300** [comment]
你好，感谢关注PaddleNLP~ 一般来说使用条件随机场可以直接使用CRF Loss，不会使用NLLoss。而且这个NLLoss的正负值跟[START]和[END]关系不大。

**Comment by shayuewt** [comment]
> 你好，感谢关注PaddleNLP~ 一般来说使用条件随机场可以直接使用CRF Loss，不会使用NLLoss。而且这个NLLoss的正负值跟[START]和[END]关系不大。

感谢您的回复。我在训练LinearChainCRF时选择的优化对象是：minimize -log(p(Y|X)) = log Z(X) - score(X, Y)。和你们的实现是同一个目标函数，可能是我没有表述对哈。 那据您所知的话，是什么可能导致最后的loss变为负值呢。前向后向算出的partition score我也有测试过，前-后向得到的结果是一致的。

**Comment by shayuewt** [comment]
晚上的时候，我特地针对不加`[START]`和`[END]`的场景，对`PaddleNlp`中计算`Partition Score`以及`Sequence Score`的模块同我自己实现的版本进行了对比，发现在多个场景下，打印的结果几乎是一致的。这是不是证明了PaddleNlp实现的版本在不加`[START]`和`[END]`的情况下可能也会出现crf loss为负数的情况呢？不知你们在测试的时候有没有实验过。下面是比较的代码。

这个问题我在网络上找了很久的答案，发现大多数开源版本的LinearChainCRF都是增加了起始与终止标识。搜到最后loss迭代为负数的案例很少，所以我真的很希望哪位同学能帮我解答一下疑惑，和我探讨一下这个起始终止符的意义。

# 计算sequence score

## PaddleNLP

```python
import paddle
import numpy as np

from paddlenlp.layers import LinearChainCrf

np.random.seed(42)
paddle.seed(42)

batch_size = 2
seq_len = 4
tag_num = 3

paddle_crf = LinearChainCrf(num_labels=tag_num, with_start_stop_tag=False)

print(paddle_crf.transitions)

"""
[[ 0.69932854,  0.65509927,  0.36617076],
[ 0.99096620, -0.59841871, -0.68482411],
[ 0.68400383,  0.96803832, -0.62446624]]
"""

unary_score = np.random.rand(batch_size, seq_len, tag_num)
unary_score = paddle.Tensor(unary_score)
batch_length = paddle.Tensor(np.asarray([seq_len, seq_len]))
targets = paddle.Tensor(np.asarray(
    [[0, 0, 1, 2],
     [1, 0, 0, 0]])
)

print(unary_score)
"""
[[[0.37454012, 0.95071431, 0.73199394],
 [0.59865848, 0.15601864, 0.15599452],
 [0.05808361, 0.86617615, 0.60111501],
 [0.70807258, 0.02058449, 0.96990985]],

[[0.83244264, 0.21233911, 0.18182497],
 [0.18340451, 0.30424224, 0.52475643],
 [0.43194502, 0.29122914, 0.61185289],
 [0.13949386, 0.29214465, 0.36636184]]]
"""

sequence_score = paddle_crf.gold_score(inputs=unary_score, labels=targets, lengths=batch_length)

print(sequence_score)

"""without mask
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [3.47888831, 3.35680566])
"""

"""with mask, true length: [4, 3]
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [3.47888831, 2.51798338])
"""

"""with mask, true length: [3, 4]
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [3.19380256, 3.35680566])
"""
```

## 个人实现

```python
import torch
import numpy as np


def _sequence_score(unary_factors, pairwise_factors, targets, mask):
    """Compute the score of specific target tag sequences.

    Args:
        unary_factors: [batch size, sequence length, tag number]
        pairwise_factors: [tag number, tag number]
        targets: [batch size, sequence length]
        mask: [batch size, sequence length], 1 stands for valid while 0 stands for invalid

    Returns:
        sequence_score: [batch size, ]
    """
    batch_size, sequence_len = targets.shape

    sequence_score = torch.zeros(batch_size)

    # Loop add two type log factors
    for k in range(sequence_len):
        # pairwise log factors
        if k != sequence_len - 1:
            pairwise_indices = targets[:, k: k + 2]
            d = pairwise_factors[pairwise_indices[:, 0], pairwise_indices[:, 1]] * mask[:, k + 1]

            sequence_score += d

        # unary log factors
        sequence_score += unary_factors[torch.arange(batch_size), k, targets[:, k]] * mask[:, k]

    return sequence_score


def main():
    # mask matrix
    mask = torch.tensor(
        [[1, 1, 1, 0],
         [1, 1, 1, 1]]
    )

    targets = torch.tensor(
        [[0, 0, 1, 2],
         [1, 0, 0, 0]]
    )

    unary_factors = torch.from_numpy(np.asarray([[[0.37454012, 0.95071431, 0.73199394],
                                                  [0.59865848, 0.15601864, 0.15599452],
                                                  [0.05808361, 0.86617615, 0.60111501],
                                                  [0.70807258, 0.02058449, 0.96990985]],

                                                 [[0.83244264, 0.21233911, 0.18182497],
                                                  [0.18340451, 0.30424224, 0.52475643],
                                                  [0.43194502, 0.29122914, 0.61185289],
                                                  [0.13949386, 0.29214465, 0.36636184]]]))

    print(unary_factors)
    pairwise_factors = torch.from_numpy(np.asarray([[0.69932854, 0.65509927, 0.36617076],
                                                    [0.99096620, -0.59841871, -0.68482411],
                                                    [0.68400383, 0.96803832, -0.62446624]]))

    sequence_score = _sequence_score(unary_factors, pairwise_factors, targets, mask)

    print(sequence_score)


"""without mask:
tensor([3.4789, 3.3568])
"""

"""with mask, true length: [4, 3]
tensor([3.4789, 2.5180])
"""

"""with mask, true length: [3, 4]
tensor([3.1938, 3.3568])
"""
main()
```

# 计算Partition Score

## PaddleNLP

```python
import paddle
import numpy as np

from paddlenlp.layers import LinearChainCrf

np.random.seed(42)
paddle.seed(42)

batch_size = 2
seq_len = 4
tag_num = 3

paddle_crf = LinearChainCrf(num_labels=tag_num, with_start_stop_tag=False)

print(paddle_crf.transitions)

"""
[[ 0.69932854,  0.65509927,  0.36617076],
[ 0.99096620, -0.59841871, -0.68482411],
[ 0.68400383,  0.96803832, -0.62446624]]
"""

unary_score = np.random.rand(batch_size, seq_len, tag_num)
unary_score = paddle.Tensor(unary_score)
batch_length = paddle.Tensor(np.asarray([seq_len - 1, seq_len]))

print(unary_score)
"""
[[[0.37454012, 0.95071431, 0.73199394],
 [0.59865848, 0.15601864, 0.15599452],
 [0.05808361, 0.86617615, 0.60111501],
 [0.70807258, 0.02058449, 0.96990985]],

[[0.83244264, 0.21233911, 0.18182497],
 [0.18340451, 0.30424224, 0.52475643],
 [0.43194502, 0.29122914, 0.61185289],
 [0.13949386, 0.29214465, 0.36636184]]]
"""

paddle_score = paddle_crf(unary_score, batch_length)

print(paddle_score)

"""without mask
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [8.08488938, 7.32290896])
"""

"""with mask, true length: [4, 3]
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [8.08488938, 5.52514496])
"""

"""with mask, true length: [3, 4]
Tensor(shape=[2], dtype=float64, place=CPUPlace, stop_gradient=False,
       [5.85468413, 7.32290896])
"""
```

## 个人实现

```python
import torch
from torch import Tensor
import numpy as np


def _forward_dp_stable_with_mask(unary_factors: Tensor, pairwise_factors: Tensor, mask: Tensor):
    """A stable implementation of forward dynamic programming algorithm supporting invariable inputs.

    Args:
        unary_factors:
        pairwise_factors:
        mask:

    Returns:
        score: The log sum exp result of partition score. [batch size,]
        alpha_table: to store the intermediate results during forward, [batch size, sequence length - 1]
    """
    batch_size, seq_len, tag_num = unary_factors.shape

    alpha_table = torch.zeros(batch_size, seq_len - 1, tag_num)

    # The whole processes of one iteration computation can be finished with matrix.
    ################################################################################

    # initialize the begin log state, which is all zeros.
    log_pre_k_state = torch.zeros(batch_size, tag_num, 1)

    for k in range(seq_len):
        """
        需要对每一步考虑 unary factor 以及 pairwise factors，围绕当前步的mask是否为0做不同判定：
        1. 如果当前步的mask为0，那么对应的 factor sum 必为 0；
        2. 若当前步的mask非0，但是下一步的mask为0，那么对应的pairwise factors为0，当前步的factor sum只需考虑 unary factor；
        """
        k_unary_factor = (unary_factors[:, k, :] * mask[:, k].view(batch_size, 1)).unsqueeze(1)

        if k != seq_len - 1:
            pairwise_factor = (pairwise_factors.unsqueeze(1) * mask[:, k + 1].unsqueeze(1)).permute(1, 0, 2)
        else:
            pairwise_factor = torch.zeros_like(pairwise_factor)

        factor_sum = k_unary_factor + pairwise_factor
        factor_sum = factor_sum * mask[:, k].reshape(-1, 1, 1)

        # get current state of position k using the state of pre-position `k - 1`
        log_k_state = torch.logsumexp(factor_sum + log_pre_k_state.permute(0, 2, 1), dim=-1, keepdim=True)

        log_k_state = (
                mask[:, k].unsqueeze(-1) * log_k_state.squeeze() +
                (1 - mask[:, k].unsqueeze(-1)) * log_pre_k_state.squeeze()
        ).unsqueeze(-1)

        # alpha_table[:, k, :] = log_k_state.squeeze()

        log_pre_k_state = log_k_state.clone()

        # DEBUG
        # print("From {}, in {}, the log_k_state is {}, the factor sum is {}".format(
        #     '_forward_dp_stable_with_mask',
        #     k,
        #     log_k_state.squeeze(),
        #     factor_sum
        # ))

    score, _ = torch.max(log_pre_k_state.squeeze(), dim=-1)

    return score, alpha_table


def main():
    # mask matrix
    mask = torch.tensor(
        [[1, 1, 1, 0],
         [1, 1, 1, 1]]
    )

    unary_factors = torch.from_numpy(np.asarray([[[0.37454012, 0.95071431, 0.73199394],
                                                  [0.59865848, 0.15601864, 0.15599452],
                                                  [0.05808361, 0.86617615, 0.60111501],
                                                  [0.70807258, 0.02058449, 0.96990985]],

                                                 [[0.83244264, 0.21233911, 0.18182497],
                                                  [0.18340451, 0.30424224, 0.52475643],
                                                  [0.43194502, 0.29122914, 0.61185289],
                                                  [0.13949386, 0.29214465, 0.36636184]]]))

    print(unary_factors)
    pairwise_factors = torch.from_numpy(np.asarray([[0.69932854, 0.65509927, 0.36617076],
                                                    [0.99096620, -0.59841871, -0.68482411],
                                                    [0.68400383, 0.96803832, -0.62446624]]))

    forward_partition_score_v2, _ = _forward_dp_stable_with_mask(unary_factors, pairwise_factors, mask)

    print(forward_partition_score_v2)


"""without mask:
tensor([8.0849, 7.3229], dtype=torch.float64)
"""

"""with mask, true length: [4, 3]
tensor([8.0849, 5.5251], dtype=torch.float64)
"""

"""with mask, true length: [3, 4]
tensor([5.8547, 7.3229], dtype=torch.float64)
"""
main()
```

**Comment by joey12300** [comment]
@shayue111 可以看看PaddleNLP LinearChainCrfLoss的实现。一般来说接近收敛时有可能出现较小负数，因为logsumexp这个函数在处理很小值的时候可能向下溢出了。PaddleNLP会在LinearChainCrfLoss实现中会加入relu保证loss为正数。

**Comment by shayuewt** [comment]
> @shayue111 可以看看PaddleNLP LinearChainCrfLoss的实现。一般来说接近收敛时有可能出现较小负数，因为logsumexp这个函数在处理很小值的时候可能向下溢出了。PaddleNLP会在LinearChainCrfLoss实现中会加入relu保证loss为正数。

感谢您的回复，加了Relu确实能够避免负loss的出现。但是我个人理解问题可能不是出现在`logsumexp`这个方法上。CRF loss的本质是：`所有可能路径的score之和`与`目标路径的score`之间的差值。如果CRF loss为负数，那是不是意味着后者的score比所有路径之和还要大？因此我感觉是某一块的实现上导致的。

**Comment by shayuewt** [comment]
我仔细对比了paddle版本与tf版本的实现，发现可能是方向搞错了。我参照tf版本的方向后，loss下降后再也不会变为负数。

具体方向出错的代码为（上述个人实现）：

```python
factor_sum = k_unary_factor + pairwise_factor
factor_sum = factor_sum * mask[:, k].reshape(-1, 1, 1)

# get current state of position k using the state of pre-position `k - 1`
log_k_state = torch.logsumexp(factor_sum + log_pre_k_state.permute(0, 2, 1), dim=-1, keepdim=True)
```

我更改到了：

```python
k_unary_factor = potential[:, k].unsqueeze(2)

factor_sum = k_unary_factor * is_k_valid.unsqueeze(-1) + pairwise_factor * is_next_k_valid.unsqueeze(-1)

# get current state of position k using the state of pre-position `k - 1`
k_state = torch.logsumexp(factor_sum + pre_k_state.unsqueeze(2), dim=1)
```
